packages/fuels-core/src/code_gen/functions_gen.rs

```diff
@@ -71,18 +75,77 @@ pub fn expand_function(
     })
 }
 
+/// Generate the `main` function of a script
+pub fn generate_script_main_function(
+    main_function_abi: &ABIFunction,
+    types: &HashMap<usize, TypeDeclaration>,
+) -> Result<TokenStream, Error> {
+    if main_function_abi.name != "main" {
+        return Err(Error::InvalidData(
+            "Script `main` function name can not be different from `main`".into(),
+        ));
+    }
+
+    let output_type_resolved = resolve_fn_output_type(main_function_abi, types)?;
+    let output_params = single_param_type_call(&output_type_resolved);
+    let output_type: TokenStream = output_type_resolved.into();
+
+    let args = function_arguments(main_function_abi, types)?;
+
+    // TODO(iqdecay): enable support for vector inputs
+    if args.iter().any(|c| c.field_type.uses_vectors()) {
+        return Err(Error::CompilationError(
+            "Script main function contains a vector in its argument types. This currently isn't supported."
+                .to_string(),
+        ));
+    }
+    let arg_names = args.iter().map(|component| &component.field_name);
+
+    let arg_declarations = args.iter().map(|component| {
+        let name = &component.field_name;
+        let field_type: TokenStream = (&component.field_type).into();
+        quote! { #name: #field_type }
+    });
+
+    let doc = expand_doc("Run the script's `main` function with the provided arguments");
+
+    let name = safe_ident("main");
+
+    Ok(quote! {
+        #doc
+        pub fn #name(&self #(,#arg_declarations)*) -> ScriptCallHandler<#output_type> {
+            let arg_name_tokens = [#(#arg_names.into_token()),*];
+            let script_data = ABIEncoder::encode(&arg_name_tokens).expect("Cannot encode script
+            arguments").resolve(0);
+            let script_binary = std::fs::read(self.binary_filepath.as_str())
+                                        .expect("Could not read from binary filepath");
+            let provider = self.wallet.get_provider().expect("Provider not set up").clone();
+            // TODO(iqdecay): handle log decoding in scripts
```

Comment:
Well, goes without saying that log decoding is a high priority, and we should address it asap! Issue created here: https://github.com/FuelLabs/fuels-rs/issues/713.